### DISCUSSION TITLE
Help with OCR

### DISCUSSION THREAD

**Discussion by jonnyl2** [koreader/koreader#10966] [Q&A] [answered]
Even though I've followed all the instructions and copied the Tesseract files into the necessary directory on my Kobo, I have never been able to get OCR to work. 

<img width="578" alt="Screenshot 2023-10-03 at 16 29 17" src="https://github.com/koreader/koreader/assets/95502269/431a514f-b580-4053-9002-fc6fecab237c">

I have only tried with English and have made sure the document language is set to English.

KOReader lets me select individual words but I'm not able to look them up in the dictionary.

Any help would be appreciated. Thanks

**Comment by Frenzie** [comment]
That looks correct at a glance. What kind of document are you trying OCR on?

**Comment by jonnyl2** [comment]
Previously I had tried it on image files of text and I thought the resolution or quality might have been too low. So I created this PDF excerpt with K2pdfopt, stripping it of its text layer, but to no avail:

[My Bondage and My Freedom - Frederick Douglass_k2opt.pdf](https://github.com/koreader/koreader/files/12795082/My.Bondage.and.My.Freedom.-.Frederick.Douglass_k2opt.pdf)

**Comment by benoit-pierre** [comment, marked as answer]
Does it work with this patch:
```diff
 frontend/apps/reader/modules/readerhighlight.lua | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git i/frontend/apps/reader/modules/readerhighlight.lua w/frontend/apps/reader/modules/readerhighlight.lua
index a8740fd00..25da2b7fd 100644
--- i/frontend/apps/reader/modules/readerhighlight.lua
+++ w/frontend/apps/reader/modules/readerhighlight.lua
@@ -1353,10 +1353,10 @@ function ReaderHighlight:lookup(selected_text, selected_link)
     end
 
     -- if we extracted text directly
-    if selected_text.text and self.hold_pos then
+    if #selected_text.text > 0 and self.hold_pos then
         self.ui:handleEvent(Event:new("LookupWord", selected_text.text, false, word_boxes, self, selected_link))
     -- or we will do OCR
-    elseif selected_text.sboxes and self.hold_pos then
+    elseif #selected_text.sboxes > 0 and self.hold_pos then
         local text = self.ui.document:getOCRText(self.hold_pos.page, selected_text.sboxes)
         if not text then
             -- getOCRText is not implemented in some document backends, but
```
?

**Reply by Frenzie** [reply]
As a quick test confirms it's still working in <https://github.com/koreader/koreader/releases/tag/v2021.10>, broken in <https://github.com/koreader/koreader/releases/tag/v2021.11>.

Edit: as well as fixed by this patch.

<img src=https://github.com/koreader/koreader/assets/202757/4ea83a06-92b4-4fe5-917d-dd77e73554a2 width=45%>

**Reply by jonnyl2** [reply]
Worked for me as well, thank you!

**Comment by jonnyl2** [comment]
A quick follow-up question: so the OCR feature is only intended for single word look-ups; it is not suitable for selecting entire phrases/paragraphs, correct?

**Reply by Frenzie** [reply]
That might be a related regression from https://github.com/koreader/koreader/commit/7a0e3d5e68e4d119cdd60060a01303380e497c74 since I think you could, though regardless whether it's possible or not, selecting more than a few words isn't a good idea. It'd take too long to OCR them to be usable.

**Reply by jonnyl2** [reply]
I was thinking it would have been amazing to be able to use OCR for chapter titles in the upcoming Custom TOC feature where no text layer is available.

**Reply by Frenzie** [reply]
Some quick testing in v2021.10 as well as [the oldest AppImage](https://github.com/koreader/koreader/releases/tag/v2018.04.10-beta) (2018.04) suggests it half-worked, meaning you could select them but it doesn't seem to do the OCR. Which is still better behavior though because that way you can still highlight, while now you can't. At a quick glance I suspect it's not allowed now because there's no text from which to continue the selection from but that'll require more investigation.

**Reply by jonnyl2** [reply]
Thanks for looking into this. It's actually still possible to select and highlight longer passages, it just doesn't update the highlight on screen until you select `Highlight` in the pop-up menu. It would be nice if OCR could be performed on the whole highlight, so it could be saved as a note in text-format, but I understand this would be quite CPU-intensive for an ereader.

**Reply by Frenzie** [reply]
The problem (partially) comes from these lines. But since those lines are 9 years old something that happens in the lines above made some assumptions untrue.

https://github.com/koreader/koreader/blob/5b5b4d9ebc970ee80d49adf66021d71af15b317c/frontend/apps/reader/modules/readerhighlight.lua#L1327-L1330
